File: ios/GSTBilling/Services/ThermalPrinter.swift
import Foundation
import CoreBluetooth

class ThermalPrinter: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
    static let shared = ThermalPrinter()
    private var centralManager: CBCentralManager?
    private var peripheral: CBPeripheral?
    private var writeCharacteristic: CBCharacteristic?
    private var onConnect: ((Bool) -> Void)?
    
    var isConnected: Bool { peripheral?.state == .connected }
    
    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }
    
    func scanAndConnect(completion: @escaping (Bool) -> Void) {
        onConnect = completion
        guard centralManager?.state == .poweredOn else { return }
        centralManager?.scanForPeripherals(withServices: nil, options: nil)
    }
    
    func disconnect() {
        if let peripheral = peripheral {
            centralManager?.cancelPeripheralConnection(peripheral)
        }
    }
    
    func printInvoice(
        businessName: String,
        invoiceNo: String,
        customerName: String,
        items: [PrintItem],
        subtotal: Double,
        tax: Double,
        total: Double,
        amountPaid: Double,
        balance: Double
    ) {
        guard let peripheral = peripheral, let characteristic = writeCharacteristic else { return }
        
        var text = centerText(businessName, width: 32) + "\n"
        text += String(repeating: "=", count: 32) + "\n"
        text += "Invoice: \(invoiceNo)\n"
        text += "Date: \(formatDate())\n"
        text += "Customer: \(customerName)\n"
        text += String(repeating: "-", count: 32) + "\n"
        text += String(format: "%-15s %4s %8s", "Item", "Qty", "Amount") + "\n"
        text += String(repeating: "-", count: 32) + "\n"
        
        for item in items {
            text += String(format: "%-15s %4d %8.2f", item.name, item.quantity, item.amount) + "\n"
        }
        
        text += String(repeating: "-", count: 32) + "\n"
        text += String(format: "%24s %8.2f", "Subtotal:", subtotal) + "\n"
        text += String(format: "%24s %8.2f", "Tax:", tax) + "\n"
        text += String(repeating: "=", count: 32) + "\n"
        text += String(format: "%24s %8.2f", "TOTAL:", total) + "\n"
        text += String(format: "%24s %8.2f", "Paid:", amountPaid) + "\n"
        text += String(format: "%24s %8.2f", "Balance:", balance) + "\n"
        text += String(repeating: "=", count: 32) + "\n"
        text += centerText("Thank you!", width: 32) + "\n\n\n"
        
        if let data = text.data(using: .ascii) {
            peripheral.writeValue(data, for: characteristic, type: .withResponse)
        }
    }
    
    func feedPaper() {
        guard let peripheral = peripheral, let characteristic = writeCharacteristic else { return }
        let cmd = Data([0x1B, 0x64, 0x03])
        peripheral.writeValue(cmd, for: characteristic, type: .withResponse)
    }
    
    private func centerText(_ text: String, width: Int) -> String {
        let padding = max(0, (width - text.count) / 2)
        return String(repeating: " ", count: padding) + text
    }
    
    private func formatDate() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd/MM/yyyy HH:mm"
        return formatter.string(from: Date())
    }
    
    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        if let name = peripheral.name, name.lowercased().contains("printer") || name.lowercased().contains("thermal") {
            self.peripheral = peripheral
            centralManager?.stopScan()
            centralManager?.connect(peripheral, options: nil)
        }
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        peripheral.delegate = self
        peripheral.discoverServices(nil)
        onConnect?(true)
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        onConnect?(false)
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let services = peripheral.services else { return }
        for service in services {
            peripheral.discoverCharacteristics(nil, for: service)
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard let characteristics = service.characteristics else { return }
        for char in characteristics {
            if char.properties.contains(.write) || char.properties.contains(.writeWithoutResponse) {
                writeCharacteristic = char
                break
            }
        }
    }
}

struct PrintItem {
    let name: String
    let quantity: Int
    let amount: Double
}

File: android/app/src/main/java/com/gstbilling/app/data/printing/ThermalPrinter.kt
package com.gstbilling.app.data.printing

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothSocket
import android.graphics.Bitmap
import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.text.SimpleDateFormat
import java.util.*

class ThermalPrinter {
    private var socket: BluetoothSocket? = null
    private var outputStream: OutputStream? = null

    fun connect(deviceAddress: String): Boolean {
        return try {
            val adapter = BluetoothAdapter.getDefaultAdapter()
            val device = adapter.getRemoteDevice(deviceAddress)
            socket = device.createRfcommSocketToServiceRecord(
                java.util.UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
            )
            socket?.connect()
            outputStream = socket?.outputStream
            true
        } catch (e: Exception) {
            false
        }
    }

    fun disconnect() {
        try {
            outputStream?.close()
            socket?.close()
        } catch (_: Exception) {}
    }

    fun printInvoice(
        businessName: String,
        invoiceNo: String,
        customerName: String,
        items: List<PrintItem>,
        subtotal: Double,
        tax: Double,
        total: Double,
        amountPaid: Double,
        balance: Double
    ) {
        val sb = StringBuilder()
        sb.appendLine(centerText(businessName, 32))
        sb.appendLine("=".repeat(32))
        sb.appendLine("Invoice: $invoiceNo")
        sb.appendLine("Date: ${SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault()).format(Date())}")
        sb.appendLine("Customer: $customerName")
        sb.appendLine("-".repeat(32))
        sb.appendLine(String.format("%-15s %4s %8s", "Item", "Qty", "Amount"))
        sb.appendLine("-".repeat(32))

        for (item in items) {
            sb.appendLine(String.format("%-15s %4d %8.2f", item.name, item.quantity, item.amount))
        }

        sb.appendLine("-".repeat(32))
        sb.appendLine(String.format("%24s %8.2f", "Subtotal:", subtotal))
        sb.appendLine(String.format("%24s %8.2f", "Tax:", tax))
        sb.appendLine("=".repeat(32))
        sb.appendLine(String.format("%24s %8.2f", "TOTAL:", total))
        sb.appendLine(String.format("%24s %8.2f", "Paid:", amountPaid))
        sb.appendLine(String.format("%24s %8.2f", "Balance:", balance))
        sb.appendLine("=".repeat(32))
        sb.appendLine(centerText("Thank you!", 32))
        sb.appendLine()
        sb.appendLine()

        outputStream?.write(sb.toString().toByteArray())
        outputStream?.flush()
    }

    fun feedPaper() {
        outputStream?.write(byteArrayOf(0x1B, 0x64, 3))
        outputStream?.flush()
    }

    private fun centerText(text: String, width: Int): String {
        val padding = ((width - text.length) / 2).coerceAtLeast(0)
        return " ".repeat(padding) + text
    }
}

data class PrintItem(
    val name: String,
    val quantity: Int,
    val amount: Double
)
